### DISCUSSION TITLE
Alternative Payment Parameter

### DISCUSSION THREAD

**Discussion by dave5105** [manuelbl/SwissQRBill#79] [General]
Hi

Im trying to make an eBill QR-Code with your generator. Ist this possible and how?

I can't figure out how to deliver the last part mentioned as (Alternatives Verfahren Parameter - AltPmt)
The SWICO structured BillInformation is correctly printed and coded, but I need this addiotional line
to make it compatible for eBills.

It's described on Page 6, chapter 3 here:
https://www.ebill.ch/dam/downloads/qr-code/d0543-de-01-spez-nutzung-alternatives-verfahren-ebill-swiss-qr-code.pdf

Would be grateful for some advise/help/examples

**Comment by manuelbl** [comment]
The SWICO bill information and the eBill information use different parts of the QR bill: SWICO information goes into the structured reference, and the eBill information goes into the alternative schemes.

I haven't tried it, but it should be possible like this:

```
bill.setAlternativeSchemes(new AlternativeScheme[]{
    new AlternativeScheme("eBill", "eBill/R/41010560425610173")
});
```

Let me know if it works.

**Reply by dave5105** [reply]
Thank you for the advise.
I tried but failed. 

I don't know whats wrong with the AlternativeSheme List. Your proposal throws errors before and when compiling.
Maybe you will figure out something with the whole code. The var Typ and eBillInfo hold the info's, "Mitteilung" contains the \\S1...
[QRdata.txt](https://github.com/manuelbl/SwissQRBill/files/9602681/QRdata.txt)


using Codecrete.SwissQRBill.Generator;
using System;
using System.Drawing;
using System.Drawing.Imaging;
using System.IO;
using System.Globalization;

namespace Codecrete.QRCode
{
    class QRCode
    {
        static void Main(string[] args)
        {

            //Testing
            //Console.WriteLine(args.Length);
            //Console.WriteLine(args.GetValue(0));

            //Pfad als Parameter einlesen
            //var Pfad = args.GetValue(0).ToString();
            var Pfad = "c:/temp/qrdata.txt";

            //Datei aus Pfad verarbeiten und in Array schreiben
            string[] Daten = File.ReadAllLines(Pfad);

            foreach (string line in Daten)
                Console.WriteLine(line);


            //Console.WriteLine(Daten[1]);

            //Parameter lesen

            var Konto = Daten[0].ToString(); //QR IBAN
            string KreditorName = Daten[1].ToString();
            string KreditorStrasse = Daten[2].ToString();
            string KreditorOrtPLZ = Daten[3].ToString();
            string KreditorLand = Daten[4].ToString();

            var Betrag = Daten[5].ToString();
            decimal BetragKonvertiert = decimal.Parse(Betrag, CultureInfo.InvariantCulture); //Trennzeichen , oder . richtig konvertieren, da PC abhängig

            var DebitorName = Daten[6].ToString();
            string DebitorStrasse = Daten[7].ToString();
            string DebitorOrtPLZ = Daten[8].ToString();
            string DebitorLand = Daten[9].ToString();

            string Referenz = Daten[10].ToString();
            string Mitteilung = Daten[11].ToString();
            string eBillInfo = Daten[12].ToString();
            string Typ = "eBill";

            /*var Konto = args.GetValue(0).ToString(); //QR IBAN
            var KreditorName = args.GetValue(1).ToString();
            var KreditorStrasse = args.GetValue(2).ToString();
            var KreditorOrtPLZ = args.GetValue(3).ToString();
            var KreditorLand = args.GetValue(4).ToString();

            var Betrag = args.GetValue(5).ToString();
            decimal BetragKonvertiert = decimal.Parse(Betrag, CultureInfo.InvariantCulture); //Trennzeichen , oder . richtig konvertieren, da PC abhängig

            var DebitorName = args.GetValue(6).ToString();
            var DebitorStrasse = args.GetValue(7).ToString();
            var DebitorOrtPLZ = args.GetValue(8).ToString();
            var DebitorLand = args.GetValue(9).ToString();

            var Referenz = args.GetValue(10).ToString();
            var Mitteilung = args.GetValue(11).ToString();
            */

            //QR Code Daten-Objekt
            Bill bill = new Bill
            {
                // Kreditor Daten               
                Account = Konto,
                Creditor = new Address
                {
                    Name = KreditorName,
                    AddressLine1 = KreditorStrasse,
                    AddressLine2 = KreditorOrtPLZ,
                    CountryCode = KreditorLand
                },

                // Betrag und Währung
                Amount = BetragKonvertiert,
                Currency = "CHF",

                // Debitor Daten
                Debtor = new Address
                {
                    Name = DebitorName,
                    AddressLine1 = DebitorStrasse,
                    AddressLine2 = DebitorOrtPLZ,
                    CountryCode = DebitorLand
                },

                // Referenz und Mitteilung
                Reference = Referenz,
                //UnstructuredMessage = Mitteilung,
                BillInformation = Mitteilung
            };

            // Formatierung
            bill.Format.OutputSize = OutputSize.QrCodeOnly;
            bill.Format.GraphicsFormat = GraphicsFormat.PNG;

            // QR Code generieren
            byte[] png = QRBill.Generate(bill);

            // Als PNG speichern
            const string path = "c:/temp/qrcode.png";
            File.WriteAllBytes(path, png);
            Console.WriteLine($"QR code saved at { Path.GetFullPath(path) }");

            // Konvertierung PNG zu BMP
            Image QRCode = Image.FromFile("c:/temp/qrcode.png");
            QRCode.Save("c:/temp/qrcode.bmp", ImageFormat.Bmp);
        }
    }
}

**Comment by manuelbl** [comment]
It's difficult to say what's wrong in your code. There is nothing related to alternative schemes. Use this code as an example:

```
        // Setup bill
        Bill bill = new Bill();
        bill.setAccount("CH4431999123000889012");
        bill.setAmountFromDouble(199.95);
        bill.setCurrency("CHF");

        // Set creditor
        Address creditor = new Address();
        creditor.setName("Robert Schneider AG");
        creditor.setAddressLine1("Rue du Lac 1268/2/22");
        creditor.setAddressLine2("2501 Biel");
        creditor.setCountryCode("CH");
        bill.setCreditor(creditor);

        // more bill data
        bill.setReference("210000000003139471430009017");
        bill.setUnstructuredMessage("Abonnement für 2020");

        // Set debtor
        Address debtor = new Address();
        debtor.setName("Pia-Maria Rutschmann-Schnyder");
        debtor.setAddressLine1("Grosse Marktgasse 28");
        debtor.setAddressLine2("9400 Rorschach");
        debtor.setCountryCode("CH");
        bill.setDebtor(debtor);

        // Swico information
        SwicoBillInformation info = new SwicoBillInformation();
        info.setInvoiceNumber("10201409");
        info.setInvoiceDate(LocalDate.of(2019, 5, 12));
        info.setCustomerReference("1400.000-53");
        info.setVatNumber("106017086");
        info.setVatDate(LocalDate.of(2018, 5, 8));
        info.setVatRate(BigDecimal.valueOf(7.7));
        bill.setBillInformation(info.encodeAsText());

        // eBill alternative scheme
        bill.setAlternativeSchemes(new AlternativeScheme[]{
                new AlternativeScheme("eBill", "eBill/R/41010560425610173")
        });

        // Set output format
        BillFormat format = bill.getFormat();
        format.setGraphicsFormat(GraphicsFormat.SVG);
        format.setOutputSize(OutputSize.QR_BILL_ONLY);
        format.setLanguage(Language.DE);

        // Generate QR bill
        byte[] svg = QRBill.generate(bill);
```

It producdes this output:

![image](https://user-images.githubusercontent.com/10954524/191181338-c3eb99e9-c00b-44ae-8433-361960303e7b.png)

If you still have problems, please provide the full error message and specify which code line it affects.

**Comment by manuelbl** [comment]
I've just realized the problem. Your using the .NET version but have started a discussion in the Java section. (I'm aware it can be confusing.)

In .NET, it looks like so:

```
Bill bill = new Bill
{
     ...

    AlternativeSchemes = new List<AlternativeScheme>(
        new[] {
            new AlternativeScheme { Name = "eBill", Instruction = "eBill/R/41010560425610173" }
    })
};
```

**Reply by dave5105** [reply]
Thank you very much for you help. I had another issue, not sure actually why, but I could solve it myself.
It's related to eBill and maybe also Postfiance specific, but the eBills are rejected as Invalid if there is an UnstructuredMessage present... 
This element/field has to be empty if the AlternativeShemes are used. *Achselzuck*

Just in case someone has something similar...

And sorry for having posted it in the wrong place. I hadn't realized it either.

**Comment by manuelbl** [comment]
I'm not aware of any such constraint but I will do my own investigations.
